File: sdk/python/src/openlit/instrumentation/openai_agents/processor.py
"""
OpenLIT OpenAI Agents Instrumentation - Native TracingProcessor Implementation

Integrates with the OpenAI Agents SDK ``TracingProcessor`` interface.
All span data fields are read at ``on_span_end`` (when fully populated).
Compliant with OTel GenAI semantic conventions (gen-ai-spans.md,
gen-ai-agent-spans.md).
"""

import time
import threading
from collections import OrderedDict
from typing import Any, TYPE_CHECKING

from opentelemetry import context as context_api
from opentelemetry.trace import SpanKind, Status, StatusCode, Link, set_span_in_context

from openlit.__helpers import (
    common_framework_span_attributes,
    handle_exception,
)
from openlit.semcov import SemanticConvention
from openlit.instrumentation.openai_agents.utils import (
    get_operation_type,
    get_span_kind,
    generate_span_name,
    is_detailed_only,
    process_span_end,
)

try:
    from agents import TracingProcessor

    if TYPE_CHECKING:
        from agents import Trace, Span
    TRACING_AVAILABLE = True
except ImportError:

    class TracingProcessor:
        """Dummy TracingProcessor for when agents is not available."""

        def force_flush(self):
            return None

        def shutdown(self):
            return None

    if TYPE_CHECKING:
        Trace = Any
        Span = Any

    TRACING_AVAILABLE = False

_MAX_HANDOFFS = 1000


class OpenLITTracingProcessor(TracingProcessor):
    """
    OpenAI Agents tracing processor that emits OTel GenAI-compliant spans.

    Thread-safety: internal dicts are protected by a lock; no shared list
    (``span_stack``) is used.
    """

    def __init__(
        self,
        tracer,
        version,
        environment,
        application_name,
        pricing_info,
        capture_message_content,
        metrics,
        disable_metrics,
        detailed_tracing,
        agent_creation_registry=None,
        **kwargs,
    ):
        super().__init__()

        self.tracer = tracer
        self.version = version
        self.environment = environment
        self.application_name = application_name
        self.pricing_info = pricing_info
        self.capture_message_content = capture_message_content
        self.metrics = metrics
        self.disable_metrics = disable_metrics
        self.detailed_tracing = detailed_tracing
        self._agent_creation_registry = agent_creation_registry

        self._lock = threading.Lock()
        # SDK span_id -> (otel_span, start_time, ctx, context_token)
        self._otel_spans: dict = {}
        # SDK trace_id -> (otel_span, start_time, ctx, context_token)
        self._root_spans: dict = {}
        # trace_id -> group_id (conversation id)
        self._trace_group_ids: dict = {}
        # Agent handoff tracker (bounded OrderedDict)
        self._handoff_tracker: OrderedDict = OrderedDict()

    # ------------------------------------------------------------------
    # Trace lifecycle
    # ------------------------------------------------------------------
    def on_trace_start(self, trace):
        try:
            trace_id = getattr(trace, "trace_id", "unknown")
            trace_name = getattr(trace, "name", "workflow")
            group_id = getattr(trace, "group_id", None)

            operation = SemanticConvention.GEN_AI_OPERATION_TYPE_FRAMEWORK
            span_name = f"{operation} {trace_name}"

            span = self.tracer.start_span(
                span_name,
                kind=SpanKind.INTERNAL,
                attributes={
                    SemanticConvention.GEN_AI_OPERATION: operation,
                    SemanticConvention.GEN_AI_PROVIDER_NAME: SemanticConvention.GEN_AI_SYSTEM_OPENAI,
                },
            )

            start_time = time.time()
            ctx = set_span_in_context(span)
            token = context_api.attach(ctx)

            with self._lock:
                self._root_spans[trace_id] = (span, start_time, ctx, token)
                if group_id:
                    self._trace_group_ids[trace_id] = str(group_id)

        except Exception as e:
            handle_exception(None, e)

    def on_trace_end(self, trace):
        try:
            trace_id = getattr(trace, "trace_id", "unknown")
            trace_name = getattr(trace, "name", "workflow")
            group_id = None

            with self._lock:
                entry = self._root_spans.pop(trace_id, None)
                group_id = self._trace_group_ids.pop(trace_id, None)

            if entry is None:
                return

            span, start_time, _ctx, token = entry
            end_time = time.time()

            scope = type("Scope", (), {})()
            scope._span = span
            scope._start_time = start_time
            scope._end_time = end_time

            common_framework_span_attributes(
                scope,
                SemanticConvention.GEN_AI_SYSTEM_OPENAI,
                "api.openai.com",
                443,
                self.environment,
                self.application_name,
                self.version,
                SemanticConvention.GEN_AI_OPERATION_TYPE_FRAMEWORK,
            )

            span.set_attribute(
                SemanticConvention.GEN_AI_OPERATION,
                SemanticConvention.GEN_AI_OPERATION_TYPE_FRAMEWORK,
            )
            span.set_attribute(
                SemanticConvention.GEN_AI_PROVIDER_NAME,
                SemanticConvention.GEN_AI_SYSTEM_OPENAI,
            )
            span.set_attribute(SemanticConvention.GEN_AI_WORKFLOW_NAME, trace_name)

            if group_id:
                span.set_attribute(SemanticConvention.GEN_AI_CONVERSATION_ID, group_id)

            error = getattr(trace, "error", None)
            if error:
                error_msg = (
                    error.get("message", "unknown")
                    if isinstance(error, dict)
                    else str(error)
                )
                span.set_attribute(SemanticConvention.ERROR_TYPE, error_msg)
                span.set_status(Status(StatusCode.ERROR, error_msg))
            else:
                span.set_status(Status(StatusCode.OK))

            if not self.disable_metrics and self.metrics:
                from openlit.instrumentation.openai_agents.utils import _record_metrics

                _record_metrics(
                    self.metrics,
                    SemanticConvention.GEN_AI_OPERATION_TYPE_FRAMEWORK,
                    end_time - start_time,
                    self.environment,
                    self.application_name,
                    None,
                    "api.openai.com",
                    443,
                )

            if token is not None:
                context_api.detach(token)

            span.end()

        except Exception as e:
            handle_exception(None, e)

    # ------------------------------------------------------------------
    # Span lifecycle
    # ------------------------------------------------------------------
    # LLM span types handled by the OpenAI SDK instrumentation, which
    # produces richer telemetry (events, cost, TTFT, streaming flags, etc.).
    _LLM_SPAN_TYPES = frozenset(("response", "generation"))

    def on_span_start(self, span):
        try:
            sdk_span = span
            span_data = sdk_span.span_data
            span_type = getattr(span_data, "type", "unknown")

            if span_type in self._LLM_SPAN_TYPES:
                return

            if is_detailed_only(span_type) and not self.detailed_tracing:
                return

            trace_id = getattr(sdk_span, "trace_id", "unknown")
            sdk_span_id = getattr(sdk_span, "span_id", None)
            parent_sdk_id = getattr(sdk_span, "parent_id", None)

            operation = get_operation_type(span_type)
            kind = get_span_kind(operation)
            span_name = generate_span_name(span_data)

            # Find parent OTel span context
            parent_ctx = None
            with self._lock:
                if parent_sdk_id and parent_sdk_id in self._otel_spans:
                    parent_otel, _, _, _ = self._otel_spans[parent_sdk_id]
                    parent_ctx = set_span_in_context(parent_otel)
                elif trace_id in self._root_spans:
                    _, _, parent_ctx, _ = self._root_spans[trace_id]

            # Span links: connect invoke_agent back to create_agent
            links = []
            if span_type == "agent" and self._agent_creation_registry:
                agent_name = getattr(span_data, "name", None)
                if agent_name:
                    creation_ctx = self._agent_creation_registry.get(str(agent_name))
                    if creation_ctx:
                        links.append(Link(creation_ctx))

            otel_span = self.tracer.start_span(
                span_name,
                kind=kind,
                context=parent_ctx,
                links=links,
                attributes={
                    SemanticConvention.GEN_AI_OPERATION: operation,
                    SemanticConvention.GEN_AI_PROVIDER_NAME: SemanticConvention.GEN_AI_SYSTEM_OPENAI,
                },
            )

            start_time = time.time()
            ctx = set_span_in_context(otel_span)
            token = context_api.attach(ctx)

            if sdk_span_id:
                with self._lock:
                    self._otel_spans[sdk_span_id] = (otel_span, start_time, ctx, token)

        except Exception as e:
            handle_exception(None, e)

    def on_span_end(self, span):
        try:
            sdk_span = span
            span_data = sdk_span.span_data
            span_type = getattr(span_data, "type", "unknown")

            if span_type in self._LLM_SPAN_TYPES:
                return

            if is_detailed_only(span_type) and not self.detailed_tracing:
                return

            sdk_span_id = getattr(sdk_span, "span_id", None)
            trace_id = getattr(sdk_span, "trace_id", "unknown")

            entry = None
            with self._lock:
                entry = self._otel_spans.pop(sdk_span_id, None) if sdk_span_id else None
                conversation_id = self._trace_group_ids.get(trace_id)

            if entry is None:
                return

            otel_span, start_time, _ctx, token = entry

            process_span_end(
                otel_span=otel_span,
                sdk_span=sdk_span,
                start_time=start_time,
                version=self.version,
                environment=self.environment,
                application_name=self.application_name,
                capture_message_content=self.capture_message_content,
                metrics=self.metrics,
                disable_metrics=self.disable_metrics,
                conversation_id=conversation_id,
                handoff_tracker=self._handoff_tracker,
            )

            if token is not None:
                context_api.detach(token)

            otel_span.end()

        except Exception as e:
            handle_exception(None, e)

    # ------------------------------------------------------------------
    # Lifecycle management
    # ------------------------------------------------------------------
    def force_flush(self):
        try:
            with self._lock:
                for otel_span, _, _, token in self._otel_spans.values():
                    try:
                        if token is not None:
                            context_api.detach(token)
                        otel_span.end()
                    except Exception:
                        pass
                self._otel_spans.clear()

                for otel_span, _, _, token in self._root_spans.values():
                    try:
                        if token is not None:
                            context_api.detach(token)
                        otel_span.end()
                    except Exception:
                        pass
                self._root_spans.clear()
                self._trace_group_ids.clear()
        except Exception as e:
            handle_exception(None, e)

    def shutdown(self):
        self.force_flush()

File: sdk/typescript/src/instrumentation/openai-agents/processor.ts
/**
 * OpenLIT OpenAI Agents TracingProcessor implementation.
 *
 * Integrates with the @openai/agents TracingProcessor interface.
 * All span data fields are read at onSpanEnd (when fully populated).
 * Compliant with OTel GenAI semantic conventions.
 */

import {
  context as contextApi,
  Span as OtelSpan,
  SpanKind,
  SpanStatusCode,
  Link,
  SpanContext,
  trace,
} from '@opentelemetry/api';
import { ATTR_SERVICE_NAME, ATTR_TELEMETRY_SDK_NAME } from '@opentelemetry/semantic-conventions';

import SemanticConvention from '../../semantic-convention';
import { SDK_NAME, SDK_VERSION } from '../../constant';
import OpenlitConfig from '../../config';
import {
  applyCustomSpanAttributes,
  getServerAddressForProvider,
  setFrameworkParentContext,
  clearFrameworkParentContext,
} from '../../helpers';
import {
  getOperationType,
  getSpanKind,
  generateSpanName,
  processSpanEnd,
  recordMetrics,
} from './utils';

const [OPENAI_SERVER_ADDRESS, OPENAI_SERVER_PORT] = getServerAddressForProvider('openai');

const LLM_SPAN_TYPES = new Set(['response', 'generation']);

interface SpanEntry {
  otelSpan: OtelSpan;
  startTime: number;
}

interface TraceEntry {
  otelSpan: OtelSpan;
  startTime: number;
}

export interface AgentCreationRegistry {
  register(agentName: string, spanContext: SpanContext): void;
  get(agentName: string): SpanContext | undefined;
}

/**
 * TracingProcessor that emits OTel GenAI-compliant spans from
 * the @openai/agents SDK tracing lifecycle.
 *
 * Thread-safe by design: each trace/span entry is keyed independently.
 * LLM span types (response, generation) are skipped -- the OpenAI
 * provider instrumentation handles those with richer telemetry.
 */
export class OpenLITTracingProcessor {
  private _tracer: ReturnType<typeof trace.getTracer>;
  private _agentCreationRegistry: AgentCreationRegistry | null;

  // SDK span_id -> SpanEntry
  private _otelSpans = new Map<string, SpanEntry>();
  // SDK trace_id -> TraceEntry
  private _rootSpans = new Map<string, TraceEntry>();
  // trace_id -> group_id (conversation id)
  private _traceGroupIds = new Map<string, string>();
  // Agent handoff tracker (bounded Map)
  private _handoffTracker = new Map<string, string>();

  constructor(
    tracer: ReturnType<typeof trace.getTracer>,
    agentCreationRegistry: AgentCreationRegistry | null = null,
  ) {
    this._tracer = tracer;
    this._agentCreationRegistry = agentCreationRegistry;
  }

  // ------------------------------------------------------------------
  // Trace lifecycle
  // ------------------------------------------------------------------
  async onTraceStart(sdkTrace: any): Promise<void> {
    try {
      const traceId: string = sdkTrace.traceId ?? 'unknown';
      const traceName: string = sdkTrace.name ?? 'workflow';
      const groupId: string | null = sdkTrace.groupId ?? null;

      const operation = SemanticConvention.GEN_AI_OPERATION_TYPE_FRAMEWORK;
      const spanName = `${operation} ${traceName}`;

      const otelSpan = this._tracer.startSpan(spanName, {
        kind: SpanKind.INTERNAL,
        attributes: {
          [SemanticConvention.GEN_AI_OPERATION]: operation,
          [SemanticConvention.GEN_AI_PROVIDER_NAME_OTEL]: SemanticConvention.GEN_AI_SYSTEM_OPENAI,
        },
      });

      const startTime = Date.now();
      this._rootSpans.set(traceId, { otelSpan, startTime });
      if (groupId) {
        this._traceGroupIds.set(traceId, String(groupId));
      }

      setFrameworkParentContext(trace.setSpan(contextApi.active(), otelSpan));
    } catch {
      // swallow
    }
  }

  async onTraceEnd(sdkTrace: any): Promise<void> {
    try {
      const traceId: string = sdkTrace.traceId ?? 'unknown';
      const traceName: string = sdkTrace.name ?? 'workflow';

      const entry = this._rootSpans.get(traceId);
      this._rootSpans.delete(traceId);
      const groupId = this._traceGroupIds.get(traceId) ?? null;
      this._traceGroupIds.delete(traceId);

      if (!entry) return;

      const { otelSpan, startTime } = entry;
      const endTime = Date.now();
      const durationMs = endTime - startTime;

      // Set common framework attributes
      otelSpan.setAttribute(ATTR_TELEMETRY_SDK_NAME, SDK_NAME);
      otelSpan.setAttribute(SemanticConvention.GEN_AI_SDK_VERSION, SDK_VERSION);
      otelSpan.setAttribute(
        SemanticConvention.GEN_AI_PROVIDER_NAME_OTEL,
        SemanticConvention.GEN_AI_SYSTEM_OPENAI,
      );
      otelSpan.setAttribute(
        SemanticConvention.GEN_AI_OPERATION,
        SemanticConvention.GEN_AI_OPERATION_TYPE_FRAMEWORK,
      );
      if (OPENAI_SERVER_ADDRESS) {
        otelSpan.setAttribute(SemanticConvention.SERVER_ADDRESS, OPENAI_SERVER_ADDRESS);
        if (OPENAI_SERVER_PORT) {
          otelSpan.setAttribute(SemanticConvention.SERVER_PORT, OPENAI_SERVER_PORT);
        }
      }
      otelSpan.setAttribute(
        SemanticConvention.ATTR_DEPLOYMENT_ENVIRONMENT,
        OpenlitConfig.environment ?? 'default',
      );
      otelSpan.setAttribute(ATTR_SERVICE_NAME, OpenlitConfig.applicationName ?? 'default');
      otelSpan.setAttribute(
        SemanticConvention.GEN_AI_CLIENT_OPERATION_DURATION,
        durationMs / 1000,
      );
      otelSpan.setAttribute(SemanticConvention.GEN_AI_WORKFLOW_NAME, traceName);

      if (groupId) {
        otelSpan.setAttribute(SemanticConvention.GEN_AI_CONVERSATION_ID, groupId);
      }

      applyCustomSpanAttributes(otelSpan);

      // Error handling
      const error = sdkTrace.error;
      if (error) {
        const errorType =
          typeof error === 'object' && error !== null
            ? (error as any).constructor?.name || (error as any).code || '_OTHER'
            : '_OTHER';
        const errorMsg =
          typeof error === 'object' && error !== null
            ? (error as any).message ?? String(error)
            : String(error);
        otelSpan.setAttribute(SemanticConvention.ERROR_TYPE, errorType);
        otelSpan.setStatus({ code: SpanStatusCode.ERROR, message: errorMsg });
      } else {
        otelSpan.setStatus({ code: SpanStatusCode.OK });
      }

      // Metrics
      if (!OpenlitConfig.disableMetrics) {
        recordMetrics(
          SemanticConvention.GEN_AI_OPERATION_TYPE_FRAMEWORK,
          durationMs / 1000,
          null,
        );
      }

      otelSpan.end();
      clearFrameworkParentContext();
    } catch {
      // swallow
    }
  }

  // ------------------------------------------------------------------
  // Span lifecycle
  // ------------------------------------------------------------------
  async onSpanStart(sdkSpan: any): Promise<void> {
    try {
      const spanData = sdkSpan.spanData;
      const spanType: string = spanData?.type ?? 'unknown';

      // Skip LLM span types -- let the OpenAI provider instrumentation handle them
      if (LLM_SPAN_TYPES.has(spanType)) return;

      const traceId: string = sdkSpan.traceId ?? 'unknown';
      const sdkSpanId: string | null = sdkSpan.spanId ?? null;
      const parentSdkId: string | null = sdkSpan.parentId ?? null;

      const operation = getOperationType(spanType);
      const kind = getSpanKind(operation);
      const spanName = generateSpanName(spanData);

      // Find parent OTel span context
      let parentCtx = contextApi.active();
      if (parentSdkId && this._otelSpans.has(parentSdkId)) {
        const parentEntry = this._otelSpans.get(parentSdkId)!;
        parentCtx = trace.setSpan(contextApi.active(), parentEntry.otelSpan);
      } else if (this._rootSpans.has(traceId)) {
        const rootEntry = this._rootSpans.get(traceId)!;
        parentCtx = trace.setSpan(contextApi.active(), rootEntry.otelSpan);
      }

      // Span links: connect invoke_agent back to create_agent
      const links: Link[] = [];
      if (spanType === 'agent' && this._agentCreationRegistry) {
        const agentName = spanData.name;
        if (agentName) {
          const creationCtx = this._agentCreationRegistry.get(String(agentName));
          if (creationCtx) {
            links.push({ context: creationCtx, attributes: {} });
          }
        }
      }

      const otelSpan = this._tracer.startSpan(
        spanName,
        {
          kind,
          links,
          attributes: {
            [SemanticConvention.GEN_AI_OPERATION]: operation,
            [SemanticConvention.GEN_AI_PROVIDER_NAME_OTEL]: SemanticConvention.GEN_AI_SYSTEM_OPENAI,
          },
        },
        parentCtx,
      );

      const startTime = Date.now();

      if (sdkSpanId) {
        this._otelSpans.set(sdkSpanId, { otelSpan, startTime });
      }

      setFrameworkParentContext(trace.setSpan(parentCtx, otelSpan));
    } catch {
      // swallow
    }
  }

  async onSpanEnd(sdkSpan: any): Promise<void> {
    try {
      const spanData = sdkSpan.spanData;
      const spanType: string = spanData?.type ?? 'unknown';

      // Skip LLM span types
      if (LLM_SPAN_TYPES.has(spanType)) return;

      const sdkSpanId: string | null = sdkSpan.spanId ?? null;
      const traceId: string = sdkSpan.traceId ?? 'unknown';

      if (!sdkSpanId) return;
      const entry = this._otelSpans.get(sdkSpanId);
      this._otelSpans.delete(sdkSpanId);
      if (!entry) return;

      const { otelSpan, startTime } = entry;
      const conversationId = this._traceGroupIds.get(traceId) ?? null;

      processSpanEnd(
        otelSpan,
        sdkSpan,
        startTime,
        conversationId,
        this._handoffTracker,
      );

      otelSpan.end();

      // Restore parent context so subsequent provider spans nest correctly
      const parentSdkId: string | null = sdkSpan.parentId ?? null;
      if (parentSdkId && this._otelSpans.has(parentSdkId)) {
        const parentEntry = this._otelSpans.get(parentSdkId)!;
        setFrameworkParentContext(trace.setSpan(contextApi.active(), parentEntry.otelSpan));
      } else if (this._rootSpans.has(traceId)) {
        const rootEntry = this._rootSpans.get(traceId)!;
        setFrameworkParentContext(trace.setSpan(contextApi.active(), rootEntry.otelSpan));
      }
    } catch {
      // swallow
    }
  }

  // ------------------------------------------------------------------
  // Lifecycle management
  // ------------------------------------------------------------------
  async forceFlush(): Promise<void> {
    try {
      for (const [, { otelSpan }] of this._otelSpans) {
        try { otelSpan.end(); } catch { /* ignore */ }
      }
      this._otelSpans.clear();

      for (const [, { otelSpan }] of this._rootSpans) {
        try { otelSpan.end(); } catch { /* ignore */ }
      }
      this._rootSpans.clear();
      this._traceGroupIds.clear();
    } catch {
      // swallow
    }
  }

  async shutdown(): Promise<void> {
    await this.forceFlush();
  }
}
